File: src/main/java/com/qwazr/jdbc/cache/ResultSetWriter.java
/*
 * Copyright 2016 Emmanuel Keller / QWAZR
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.qwazr.jdbc.cache;

import java.io.ByteArrayOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.math.BigDecimal;
import java.nio.file.Path;
import java.sql.Clob;
import java.sql.Date;
import java.sql.ResultSet;
import java.sql.ResultSetMetaData;
import java.sql.RowId;
import java.sql.SQLException;
import java.sql.Time;
import java.sql.Timestamp;
import java.sql.Types;
import java.util.zip.GZIPOutputStream;

class ResultSetWriter {

    static void write(final Path resultSetPath, final ResultSet resultSet) throws SQLException {
        try (final FileOutputStream fos = new FileOutputStream(resultSetPath.toFile())) {
            try (final GZIPOutputStream zos = new GZIPOutputStream(fos)) {
                try (final DataOutputStream output = new DataOutputStream(zos)) {
                    writeMetadata(output, resultSet.getMetaData());
                    writeResultSet(output, resultSet);
                }
            }
        } catch (IOException e) {
            throw new SQLException("Error while writing the ResultSet cache file: " + resultSetPath, e);
        }
    }

    static ByteArrayOutputStream write(final ResultSet resultSet) throws SQLException {
        try (final ByteArrayOutputStream fos = new ByteArrayOutputStream()) {
            try (final DataOutputStream output = new DataOutputStream(fos)) {
                writeMetadata(output, resultSet.getMetaData());
                writeResultSet(output, resultSet);
                return fos;
            }
        } catch (IOException e) {
            throw new SQLException("Error while writing the ResultSet cache", e);
        }
    }

    private static void writeMetadata(final DataOutputStream output, final ResultSetMetaData metadata)
            throws IOException, SQLException {
        final int columnCount = metadata.getColumnCount();
        output.writeInt(columnCount);
        for (int i = 1; i <= columnCount; i++) {
            output.writeUTF(metadata.getCatalogName(i));
            output.writeUTF(metadata.getColumnClassName(i));
            output.writeUTF(metadata.getColumnLabel(i));
            output.writeUTF(metadata.getColumnName(i));
            output.writeUTF(metadata.getColumnTypeName(i));
            output.writeInt(metadata.getColumnType(i));
            output.writeInt(metadata.getColumnDisplaySize(i));
            output.writeInt(metadata.getPrecision(i));
            output.writeUTF(metadata.getTableName(i));
            output.writeInt(metadata.getScale(i));
            output.writeUTF(metadata.getSchemaName(i));
            output.writeBoolean(metadata.isAutoIncrement(i));
            output.writeBoolean(metadata.isCaseSensitive(i));
            output.writeBoolean(metadata.isCurrency(i));
            output.writeBoolean(metadata.isDefinitelyWritable(i));
            output.writeInt(metadata.isNullable(i));
            output.writeBoolean(metadata.isReadOnly(i));
            output.writeBoolean(metadata.isSearchable(i));
            output.writeBoolean(metadata.isSigned(i));
            output.writeBoolean(metadata.isWritable(i));
        }

    }

    static ColumnDef[] readColumns(final DataInputStream input) throws IOException {
        final int columnCount = input.readInt();
        final ColumnDef[] columns = new ColumnDef[columnCount];
        for (int i = 0; i < columnCount; i++)
            columns[i] = new ColumnDef(input);
        return columns;
    }

    final static class ColumnDef {

        final String catalog;
        final String className;
        final String label;
        final String name;
        final String typeName;
        final int type;
        final int displaySize;
        final int precision;
        final String tableName;
        final int scale;
        final String schemaName;
        final boolean isAutoIncrement;
        final boolean isCaseSensitive;
        final boolean isCurrency;
        final boolean isDefinitelyWritable;
        final int isNullable;
        final boolean isReadOnly;
        final boolean isSearchable;
        final boolean isSigned;
        final boolean isWritable;

        private ColumnDef(final DataInputStream input) throws IOException {
            catalog = input.readUTF();
            className = input.readUTF();
            label = input.readUTF();
            name = input.readUTF();
            typeName = input.readUTF();
            type = input.readInt();
            displaySize = input.readInt();
            precision = input.readInt();
            tableName = input.readUTF();
            scale = input.readInt();
            schemaName = input.readUTF();
            isAutoIncrement = input.readBoolean();
            isCaseSensitive = input.readBoolean();
            isCurrency = input.readBoolean();
            isDefinitelyWritable = input.readBoolean();
            isNullable = input.readInt();
            isReadOnly = input.readBoolean();
            isSearchable = input.readBoolean();
            isSigned = input.readBoolean();
            isWritable = input.readBoolean();
        }
    }

    private static void writeResultSet(final DataOutputStream output, final ResultSet resultSet)
            throws SQLException, IOException {
        final ResultSetMetaData metaData = resultSet.getMetaData();
        final int[] types = new int[metaData.getColumnCount()];
        for (int i = 0; i < types.length; i++)
            types[i] = metaData.getColumnType(i + 1);
        int pos = 0;
        while (resultSet.next()) {
            output.writeInt(++pos);
            int i = 0;
            for (int type : types) {
                i++;
                switch (type) {
                case Types.BIT:
                case Types.BOOLEAN:
                    writeBoolean(i, resultSet, output);
                    break;
                case Types.TINYINT:
                    writeByte(i, resultSet, output);
                    break;
                case Types.SMALLINT:
                    writeShort(i, resultSet, output);
                    break;
                case Types.INTEGER:
                    writeInteger(i, resultSet, output);
                    break;
                case Types.BIGINT:
                    writeLong(i, resultSet, output);
                    break;
                case Types.FLOAT:
                case Types.REAL:
                    writeFloat(i, resultSet, output);
                    break;
                case Types.DOUBLE:
                    writeDouble(i, resultSet, output);
                    break;
                case Types.NUMERIC:
                case Types.DECIMAL:
                    writeBigDecimal(i, resultSet, output);
                    break;
                case Types.CHAR:
                case Types.VARCHAR:
                case Types.LONGVARCHAR:
                case Types.NCHAR:
                case Types.NVARCHAR:
                case Types.LONGNVARCHAR:
                    writeString(i, resultSet, output);
                    break;
                case Types.DATE:
                    writeDate(i, resultSet, output);
                    break;
                case Types.TIME:
                case Types.TIME_WITH_TIMEZONE:
                    writeTime(i, resultSet, output);
                    break;
                case Types.TIMESTAMP:
                case Types.TIMESTAMP_WITH_TIMEZONE:
                    writeTimestamp(i, resultSet, output);
                    break;
                case Types.ROWID:
                    writeRowId(i, resultSet, output);
                    break;
                case Types.CLOB:
                    writeClob(i, resultSet, output);
                    break;
                case Types.BINARY:
                case Types.VARBINARY:
                case Types.LONGVARBINARY:
                case Types.NULL:
                case Types.OTHER:
                case Types.JAVA_OBJECT:
                case Types.DISTINCT:
                case Types.STRUCT:
                case Types.ARRAY:
                case Types.BLOB:
                case Types.REF:
                case Types.DATALINK:
                case Types.NCLOB:
                case Types.SQLXML:
                case Types.REF_CURSOR:
                    writeNull(output);
                    break;
                }
            }
        }
    }

    private static void writeBoolean(final int column, final ResultSet resultSet, final DataOutputStream output)
            throws SQLException, IOException {
        final boolean val = resultSet.getBoolean(column);
        final boolean wasNull = resultSet.wasNull();
        output.writeBoolean(!wasNull);
        if (!wasNull)
            output.writeBoolean(val);
    }

    private static void writeByte(final int column, final ResultSet resultSet, final DataOutputStream output)
            throws SQLException, IOException {
        final byte val = resultSet.getByte(column);
        final boolean wasNull = resultSet.wasNull();
        output.writeBoolean(!wasNull);
        if (!wasNull)
            output.writeByte(val);
    }

    private static void writeShort(final int column, final ResultSet resultSet, final DataOutputStream output)
            throws SQLException, IOException {
        final short val = resultSet.getShort(column);
        final boolean wasNull = resultSet.wasNull();
        output.writeBoolean(!wasNull);
        if (!wasNull)
            output.writeShort(val);
    }

    private static void writeInteger(final int column, final ResultSet resultSet, final DataOutputStream output)
            throws SQLException, IOException {
        final int val = resultSet.getInt(column);
        final boolean wasNull = resultSet.wasNull();
        output.writeBoolean(!wasNull);
        if (!wasNull)
            output.writeInt(val);
    }

    private static void writeLong(final int column, final ResultSet resultSet, final DataOutputStream output)
            throws SQLException, IOException {
        final long val = resultSet.getLong(column);
        final boolean wasNull = resultSet.wasNull();
        output.writeBoolean(!wasNull);
        if (!wasNull)
            output.writeLong(val);
    }

    private static void writeFloat(final int column, final ResultSet resultSet, final DataOutputStream output)
            throws SQLException, IOException {
        final float val = resultSet.getFloat(column);
        final boolean wasNull = resultSet.wasNull();
        output.writeBoolean(!wasNull);
        if (!wasNull)
            output.writeFloat(val);
    }

    private static void writeDouble(final int column, final ResultSet resultSet, final DataOutputStream output)
            throws SQLException, IOException {
        final double val = resultSet.getDouble(column);
        final boolean wasNull = resultSet.wasNull();
        output.writeBoolean(!wasNull);
        if (!wasNull)
            output.writeDouble(val);
    }

    private static void writeString(final int column, final ResultSet resultSet, final DataOutputStream output)
            throws SQLException, IOException {
        final String val = resultSet.getString(column);
        final boolean wasNull = resultSet.wasNull();
        output.writeBoolean(!wasNull);
        if (!wasNull)
            output.writeUTF(val);
    }

    private static void writeBigDecimal(final int column, final ResultSet resultSet, final DataOutputStream output)
            throws SQLException, IOException {
        final BigDecimal val = resultSet.getBigDecimal(column);
        final boolean wasNull = resultSet.wasNull();
        output.writeBoolean(!wasNull);
        if (wasNull)
            return;
        output.writeDouble(val.doubleValue());
    }

    private static void writeDate(final int column, final ResultSet resultSet, final DataOutputStream output)
            throws SQLException, IOException {
        final Date val = resultSet.getDate(column);
        final boolean wasNull = resultSet.wasNull();
        output.writeBoolean(!wasNull);
        if (!wasNull)
            output.writeLong(val.getTime());
    }

    private static void writeTime(final int column, final ResultSet resultSet, final DataOutputStream output)
            throws SQLException, IOException {
        final Time val = resultSet.getTime(column);
        final boolean wasNull = resultSet.wasNull();
        output.writeBoolean(!wasNull);
        if (!wasNull)
            output.writeLong(val.getTime());
    }

    private static void writeTimestamp(final int column, final ResultSet resultSet, final DataOutputStream output)
            throws SQLException, IOException {
        final Timestamp val = resultSet.getTimestamp(column);
        final boolean wasNull = resultSet.wasNull();
        output.writeBoolean(!wasNull);
        if (!wasNull)
            output.writeLong(val.getTime());
    }

    private static void writeRowId(final int column, final ResultSet resultSet, final DataOutputStream output)
            throws SQLException, IOException {
        final RowId val = resultSet.getRowId(column);
        final boolean wasNull = resultSet.wasNull();
        output.writeBoolean(!wasNull);
        if (!wasNull)
            output.writeUTF(val.toString());
    }

    private static void writeClob(final int column, final ResultSet resultSet, final DataOutputStream output)
            throws SQLException, IOException {
        final Clob clob = resultSet.getClob(column);
        try {
            final boolean wasNull = resultSet.wasNull();
            output.writeBoolean(!wasNull);
            if (wasNull)
                return;
            // TODO Support reader for long sized string
            final long size = clob.length();
            output.writeUTF(size == 0 ? "" : clob.getSubString(1, (int) clob.length()));
        } finally {
            if (clob != null) {
                try {
                    clob.free();
                } catch (AbstractMethodError e) {
                    // May occur with old JDBC drivers
                }
            }
        }
    }

    private static void writeNull(final DataOutputStream output) throws IOException {
        output.writeBoolean(false);
    }

    public static Object readRow(final int type, final DataInputStream input) throws IOException {
        final boolean wasNull = !input.readBoolean();
        if (wasNull)
            return null;
        switch (type) {
        case Types.BIT:
        case Types.BOOLEAN:
            return input.readBoolean();
        case Types.TINYINT:
            return input.readByte();
        case Types.SMALLINT:
            return input.readShort();
        case Types.INTEGER:
            return input.readInt();
        case Types.BIGINT:
            return input.readLong();
        case Types.FLOAT:
        case Types.REAL:
            return input.readFloat();
        case Types.DOUBLE:
        case Types.NUMERIC:
        case Types.DECIMAL:
            return input.readDouble();
        case Types.CHAR:
        case Types.VARCHAR:
        case Types.LONGVARCHAR:
        case Types.NCHAR:
        case Types.NVARCHAR:
        case Types.LONGNVARCHAR:
            return input.readUTF();
        case Types.DATE:
            return new java.sql.Date(input.readLong());
        case Types.TIME:
        case Types.TIME_WITH_TIMEZONE:
            return new java.sql.Time(input.readLong());
        case Types.TIMESTAMP:
        case Types.TIMESTAMP_WITH_TIMEZONE:
            return new java.sql.Timestamp(input.readLong());
        case Types.ROWID:
            return input.readUTF();
        case Types.CLOB:
            return input.readUTF();
        case Types.BINARY:
        case Types.VARBINARY:
        case Types.LONGVARBINARY:
        case Types.NULL:
        case Types.OTHER:
        case Types.JAVA_OBJECT:
        case Types.DISTINCT:
        case Types.STRUCT:
        case Types.ARRAY:
        case Types.BLOB:
        case Types.REF:
        case Types.DATALINK:
        case Types.NCLOB:
        case Types.SQLXML:
        case Types.REF_CURSOR:
        default:
            throw new IOException("Column type no supported: " + type);
        }
    }
}

File: src/main/kotlin/io/github/jhstatewide/jdbc/cache/ResultSetWriter.kt
/*
 * Copyright 2016 Emmanuel Keller / QWAZR
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */package io.github.jhstatewide.jdbc.cache

import java.io.*
import java.nio.file.Path
import java.sql.*
import java.util.zip.GZIPOutputStream

internal object ResultSetWriter {
    @Throws(SQLException::class)
    fun write(resultSetPath: Path, resultSet: ResultSet) {
        try {
            FileOutputStream(resultSetPath.toFile()).use { fos ->
                GZIPOutputStream(fos).use { zos ->
                    DataOutputStream(zos).use { output ->
                        writeMetadata(output, resultSet.metaData)
                        writeResultSet(output, resultSet)
                    }
                }
            }
        } catch (e: IOException) {
            throw SQLException("Error while writing the ResultSet cache file: $resultSetPath", e)
        }
    }

    @Throws(SQLException::class)
    fun write(resultSet: ResultSet): ByteArrayOutputStream {
        try {
            ByteArrayOutputStream().use { fos ->
                DataOutputStream(fos).use { output ->
                    writeMetadata(output, resultSet.metaData)
                    writeResultSet(output, resultSet)
                    return fos
                }
            }
        } catch (e: IOException) {
            throw SQLException("Error while writing the ResultSet cache", e)
        }
    }

    @Throws(IOException::class, SQLException::class)
    private fun writeMetadata(output: DataOutputStream, metadata: ResultSetMetaData) {
        val columnCount = metadata.columnCount
        output.writeInt(columnCount)
        for (i in 1..columnCount) {
            output.writeUTF(metadata.getCatalogName(i))
            output.writeUTF(metadata.getColumnClassName(i))
            output.writeUTF(metadata.getColumnLabel(i))
            output.writeUTF(metadata.getColumnName(i))
            output.writeUTF(metadata.getColumnTypeName(i))
            output.writeInt(metadata.getColumnType(i))
            output.writeInt(metadata.getColumnDisplaySize(i))
            output.writeInt(metadata.getPrecision(i))
            output.writeUTF(metadata.getTableName(i))
            output.writeInt(metadata.getScale(i))
            output.writeUTF(metadata.getSchemaName(i))
            output.writeBoolean(metadata.isAutoIncrement(i))
            output.writeBoolean(metadata.isCaseSensitive(i))
            output.writeBoolean(metadata.isCurrency(i))
            output.writeBoolean(metadata.isDefinitelyWritable(i))
            output.writeInt(metadata.isNullable(i))
            output.writeBoolean(metadata.isReadOnly(i))
            output.writeBoolean(metadata.isSearchable(i))
            output.writeBoolean(metadata.isSigned(i))
            output.writeBoolean(metadata.isWritable(i))
        }
    }

    @JvmStatic
    @Throws(IOException::class)
    fun readColumns(input: DataInputStream): Array<ColumnDef> {
        val columnCount = input.readInt()
        return Array(columnCount) { ColumnDef(input) }
    }

    @Throws(SQLException::class, IOException::class)
    private fun writeResultSet(output: DataOutputStream, resultSet: ResultSet) {
        val metaData = resultSet.metaData
        val types = IntArray(metaData.columnCount)
        for (i in types.indices) types[i] = metaData.getColumnType(i + 1)
        var pos = 0
        while (resultSet.next()) {
            output.writeInt(++pos)
            var i = 0
            for (type in types) {
                i++
                when (type) {
                    Types.BIT, Types.BOOLEAN -> writeBoolean(i, resultSet, output)
                    Types.TINYINT -> writeByte(i, resultSet, output)
                    Types.SMALLINT -> writeShort(i, resultSet, output)
                    Types.INTEGER -> writeInteger(i, resultSet, output)
                    Types.BIGINT -> writeLong(i, resultSet, output)
                    Types.FLOAT, Types.REAL -> writeFloat(i, resultSet, output)
                    Types.DOUBLE -> writeDouble(i, resultSet, output)
                    Types.NUMERIC, Types.DECIMAL -> writeBigDecimal(i, resultSet, output)
                    Types.CHAR, Types.VARCHAR, Types.LONGVARCHAR, Types.NCHAR, Types.NVARCHAR, Types.LONGNVARCHAR -> writeString(
                        i,
                        resultSet,
                        output
                    )

                    Types.DATE -> writeDate(i, resultSet, output)
                    Types.TIME, Types.TIME_WITH_TIMEZONE -> writeTime(i, resultSet, output)
                    Types.TIMESTAMP, Types.TIMESTAMP_WITH_TIMEZONE -> writeTimestamp(i, resultSet, output)
                    Types.ROWID -> writeRowId(i, resultSet, output)
                    Types.CLOB -> writeClob(i, resultSet, output)
                    Types.BINARY, Types.VARBINARY, Types.LONGVARBINARY, Types.NULL, Types.OTHER, Types.JAVA_OBJECT, Types.DISTINCT, Types.STRUCT, Types.ARRAY, Types.BLOB, Types.REF, Types.DATALINK, Types.NCLOB, Types.SQLXML, Types.REF_CURSOR -> writeNull(
                        output
                    )
                }
            }
        }
    }

    @Throws(SQLException::class, IOException::class)
    private fun writeBoolean(column: Int, resultSet: ResultSet, output: DataOutputStream) {
        val `val` = resultSet.getBoolean(column)
        val wasNull = resultSet.wasNull()
        output.writeBoolean(!wasNull)
        if (!wasNull) output.writeBoolean(`val`)
    }

    @Throws(SQLException::class, IOException::class)
    private fun writeByte(column: Int, resultSet: ResultSet, output: DataOutputStream) {
        val `val` = resultSet.getByte(column)
        val wasNull = resultSet.wasNull()
        output.writeBoolean(!wasNull)
        if (!wasNull) output.writeByte(`val`.toInt())
    }

    @Throws(SQLException::class, IOException::class)
    private fun writeShort(column: Int, resultSet: ResultSet, output: DataOutputStream) {
        val `val` = resultSet.getShort(column)
        val wasNull = resultSet.wasNull()
        output.writeBoolean(!wasNull)
        if (!wasNull) output.writeShort(`val`.toInt())
    }

    @Throws(SQLException::class, IOException::class)
    private fun writeInteger(column: Int, resultSet: ResultSet, output: DataOutputStream) {
        val `val` = resultSet.getInt(column)
        val wasNull = resultSet.wasNull()
        output.writeBoolean(!wasNull)
        if (!wasNull) output.writeInt(`val`)
    }

    @Throws(SQLException::class, IOException::class)
    private fun writeLong(column: Int, resultSet: ResultSet, output: DataOutputStream) {
        val `val` = resultSet.getLong(column)
        val wasNull = resultSet.wasNull()
        output.writeBoolean(!wasNull)
        if (!wasNull) output.writeLong(`val`)
    }

    @Throws(SQLException::class, IOException::class)
    private fun writeFloat(column: Int, resultSet: ResultSet, output: DataOutputStream) {
        val `val` = resultSet.getFloat(column)
        val wasNull = resultSet.wasNull()
        output.writeBoolean(!wasNull)
        if (!wasNull) output.writeFloat(`val`)
    }

    @Throws(SQLException::class, IOException::class)
    private fun writeDouble(column: Int, resultSet: ResultSet, output: DataOutputStream) {
        val `val` = resultSet.getDouble(column)
        val wasNull = resultSet.wasNull()
        output.writeBoolean(!wasNull)
        if (!wasNull) output.writeDouble(`val`)
    }

    @Throws(SQLException::class, IOException::class)
    private fun writeString(column: Int, resultSet: ResultSet, output: DataOutputStream) {
        val `val` = resultSet.getString(column)
        val wasNull = resultSet.wasNull()
        output.writeBoolean(!wasNull)
        if (!wasNull) output.writeUTF(`val`)
    }

    @Throws(SQLException::class, IOException::class)
    private fun writeBigDecimal(column: Int, resultSet: ResultSet, output: DataOutputStream) {
        val `val` = resultSet.getBigDecimal(column)
        val wasNull = resultSet.wasNull()
        output.writeBoolean(!wasNull)
        if (wasNull) return
        output.writeDouble(`val`.toDouble())
    }

    @Throws(SQLException::class, IOException::class)
    private fun writeDate(column: Int, resultSet: ResultSet, output: DataOutputStream) {
        val `val` = resultSet.getDate(column)
        val wasNull = resultSet.wasNull()
        output.writeBoolean(!wasNull)
        if (!wasNull) output.writeLong(`val`.time)
    }

    @Throws(SQLException::class, IOException::class)
    private fun writeTime(column: Int, resultSet: ResultSet, output: DataOutputStream) {
        val `val` = resultSet.getTime(column)
        val wasNull = resultSet.wasNull()
        output.writeBoolean(!wasNull)
        if (!wasNull) output.writeLong(`val`.time)
    }

    @Throws(SQLException::class, IOException::class)
    private fun writeTimestamp(column: Int, resultSet: ResultSet, output: DataOutputStream) {
        val `val` = resultSet.getTimestamp(column)
        val wasNull = resultSet.wasNull()
        output.writeBoolean(!wasNull)
        if (!wasNull) output.writeLong(`val`.time)
    }

    @Throws(SQLException::class, IOException::class)
    private fun writeRowId(column: Int, resultSet: ResultSet, output: DataOutputStream) {
        val `val` = resultSet.getRowId(column)
        val wasNull = resultSet.wasNull()
        output.writeBoolean(!wasNull)
        if (!wasNull) output.writeUTF(`val`.toString())
    }

    @Throws(SQLException::class, IOException::class)
    private fun writeClob(column: Int, resultSet: ResultSet, output: DataOutputStream) {
        val clob = resultSet.getClob(column)
        try {
            val wasNull = resultSet.wasNull()
            output.writeBoolean(!wasNull)
            if (wasNull) return
            // TODO Support reader for long sized string
            val size = clob!!.length()
            output.writeUTF(if (size == 0L) "" else clob.getSubString(1, clob.length().toInt()))
        } finally {
            if (clob != null) {
                try {
                    clob.free()
                } catch (e: AbstractMethodError) {
                    // May occur with old JDBC drivers
                }
            }
        }
    }

    @Throws(IOException::class)
    private fun writeNull(output: DataOutputStream) {
        output.writeBoolean(false)
    }

    @JvmStatic
    @Throws(IOException::class)
    fun readRow(type: Int, input: DataInputStream): Any? {
        val wasNull = !input.readBoolean()
        return if (wasNull) null else when (type) {
            Types.BIT, Types.BOOLEAN -> input.readBoolean()
            Types.TINYINT -> input.readByte()
            Types.SMALLINT -> input.readShort()
            Types.INTEGER -> input.readInt()
            Types.BIGINT -> input.readLong()
            Types.FLOAT, Types.REAL -> input.readFloat()
            Types.DOUBLE, Types.NUMERIC, Types.DECIMAL -> input.readDouble()
            Types.CHAR, Types.VARCHAR, Types.LONGVARCHAR, Types.NCHAR, Types.NVARCHAR, Types.LONGNVARCHAR -> input.readUTF()
            Types.DATE -> Date(input.readLong())
            Types.TIME, Types.TIME_WITH_TIMEZONE -> Time(input.readLong())
            Types.TIMESTAMP, Types.TIMESTAMP_WITH_TIMEZONE -> Timestamp(input.readLong())
            Types.ROWID -> input.readUTF()
            Types.CLOB -> input.readUTF()
            Types.BINARY, Types.VARBINARY, Types.LONGVARBINARY, Types.NULL, Types.OTHER, Types.JAVA_OBJECT, Types.DISTINCT, Types.STRUCT, Types.ARRAY, Types.BLOB, Types.REF, Types.DATALINK, Types.NCLOB, Types.SQLXML, Types.REF_CURSOR -> throw IOException(
                "Column type no supported: $type"
            )

            else -> throw IOException("Column type no supported: $type")
        }
    }

    internal class ColumnDef constructor(input: DataInputStream) {
        val catalog: String
        val className: String
        @JvmField
        val label: String
        val name: String
        val typeName: String
        @JvmField
        val type: Int
        val displaySize: Int
        val precision: Int
        val tableName: String
        val scale: Int
        val schemaName: String
        val isAutoIncrement: Boolean
        val isCaseSensitive: Boolean
        val isCurrency: Boolean
        val isDefinitelyWritable: Boolean
        val isNullable: Int
        val isReadOnly: Boolean
        val isSearchable: Boolean
        val isSigned: Boolean
        val isWritable: Boolean

        init {
            catalog = input.readUTF()
            className = input.readUTF()
            label = input.readUTF()
            name = input.readUTF()
            typeName = input.readUTF()
            type = input.readInt()
            displaySize = input.readInt()
            precision = input.readInt()
            tableName = input.readUTF()
            scale = input.readInt()
            schemaName = input.readUTF()
            isAutoIncrement = input.readBoolean()
            isCaseSensitive = input.readBoolean()
            isCurrency = input.readBoolean()
            isDefinitelyWritable = input.readBoolean()
            isNullable = input.readInt()
            isReadOnly = input.readBoolean()
            isSearchable = input.readBoolean()
            isSigned = input.readBoolean()
            isWritable = input.readBoolean()
        }
    }
}